File: backend/scripts/dev.py
from __future__ import annotations

import signal
import subprocess
import sys
import time
from pathlib import Path


BACKEND_DIR = Path(__file__).resolve().parents[1]
REPO_DIR = BACKEND_DIR.parent


def main() -> int:
    run([sys.executable, "-m", "alembic", "upgrade", "head"], cwd=BACKEND_DIR)
    run([sys.executable, "scripts/seed.py"], cwd=BACKEND_DIR)

    processes = [
        subprocess.Popen(
            [sys.executable, "-m", "uvicorn", "app.main:app", "--reload", "--host", "127.0.0.1", "--port", "8000"],
            cwd=BACKEND_DIR,
        ),
        subprocess.Popen(["npm", "--prefix", str(REPO_DIR / "frontend"), "run", "dev"], cwd=REPO_DIR),
    ]

    stopping = False

    def stop(_: int | None = None, __: object | None = None) -> None:
        nonlocal stopping
        stopping = True
        for process in processes:
            if process.poll() is None:
                process.terminate()

    signal.signal(signal.SIGINT, stop)
    signal.signal(signal.SIGTERM, stop)

    try:
        while True:
            for process in processes:
                code = process.poll()
                if code is not None:
                    stop()
                    return 0 if stopping else code
            time.sleep(0.5)
    finally:
        stop()


def run(command: list[str], cwd: Path) -> None:
    result = subprocess.run(command, cwd=cwd, check=False)
    if result.returncode != 0:
        raise SystemExit(result.returncode)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/dev.mjs
#!/usr/bin/env node

import { spawn, spawnSync } from "node:child_process";

run("npm", ["run", "mock:db:push"]);
run("npm", ["run", "mock:db:seed"]);

const processes = [
  spawn("npm", ["run", "mock:api"], { stdio: "inherit" }),
  spawn("npm", ["run", "dev:web"], { stdio: "inherit" }),
];

let stopping = false;

for (const child of processes) {
  child.on("exit", (code) => {
    if (!stopping && code !== 0) {
      stop();
      process.exit(code ?? 1);
    }
  });
}

process.on("SIGINT", stop);
process.on("SIGTERM", stop);

function run(command, args) {
  const result = spawnSync(command, args, { stdio: "inherit" });
  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }
}

function stop() {
  stopping = true;
  for (const child of processes) {
    if (!child.killed) child.kill("SIGTERM");
  }
}
